### PR TITLE
[WIP] New function Reader::read() to read specific types from Input

### PR DESCRIPTION
**Pull request by oherrala**
This pull request consists of the following parts:

1. A concept of endianess (aka byte order) for both `Input` and `Reader` structs. `Input::from_be()` and `Input::from_le()` can be used to construct `Input` with big or little endian awareness. `Input::from()` expects big endian.
1. New method `Reader::read()` which can be used to read any type implementing `FromReader` trait. Reads happen in byte order specified when constructing `Input`.
1. New trait `FromReader` with two methods: `read_be()` and `read_le()`. Implementations of this trait takes care of reading a type from `Reader` in either big or little endian byte order.
1. `FromReader` implementations for Rust's standard unsigned and signed integer types.
1. Quickcheck properties for testing `Reader::read()` and `FromReader` implementations.

Things not yet implemented, but worth considering:

* It's not possible to change endianess of `Input` returned by `Reader`s methods `get_input_between_marks()`, `skip_and_get_input()` and `skip_to_end()`.
* `FromReader` implementations for some (obvious?) types like tuples and arrays.

I agree to license my contributions to each file under the terms given at the top of each file I changed.

### CODE REVIEW

**Comment by briansmith** [comment]
> I think this is also a good time to look at other crates that might already have subsumed what `untrusted` does, where it would make more sense to switch _ring_ and webpki to that other crate. I think there are a lot of alternatives now. WDYT?

I did look into some other crates that are doing similar things. Some of them are definitely more powerful than `untrusted`. However, I still prefer to use `untrusted` for my own use cases because it is so much simpler than those other crates, especially after the most recent changes and the pending #28.

I would definitely like to have functionality like this in `untrusted` if you are still interested in updating this PR.

**Comment by oherrala** [comment]
@briansmith I'm myself using my [`untrustended`](https://docs.rs/untrustended/) crate to get this functionality. Would be awesome to get this integrated so I'm still interested to work with this.

File: Cargo.toml [hunk 1]
```diff
@@ -26,3 +26,8 @@ rpath = false
 lto = true
 debug-assertions = false
 codegen-units = 1
+
+[dev-dependencies]
+byteorder = "1.1.0"
+quickcheck = { version = "0.4.1", default_features = false }
+rand = "0.3"
```
[the remark this answers is not in the document]
Comment (oherrala) [reply Cargo.toml]: I removed the whole benchmark suite (commit 2a7a1650ff2b6c3ff52a7d1e84818e027e868228). Maybe we should rewrite it in criterion instead.

**Comment by avitex** [comment]
Any work that is needed to get this through? :)

**Comment by oherrala** [comment]
@avitex I'm still interested about getting this in untrusted. In the meantime I'm using my own crate (https://github.com/oherrala/untrustended) to get the similar functionality.

I think we need some input from @briansmith if there's still interested in this PR and what needs to be done to get it merged.

**Comment by briansmith** [comment]
Note: I renamed the "master" branch to "main". Sorry for the inconvenience. This PR has had its base branch updated to "main" but you'll need to deal with the change in your local repo yourself.

**Comment by briansmith** [comment]
I'm going to close this because it's too stale. Sorry I let you down with the slow response back when you submitted this! I appreciate the effort.

**Pull request closed without merging**